drivers/CpuUid.h

```diff
@@ -0,0 +1,106 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_CPUUID_H
+#define MBED_CPUUID_H
+
+#include "platform/platform.h"
+
+#if defined(DEVICE_CPUUID) || defined(DOXYGEN_ONLY)
+
+#include <string>
+#include <vector>
+
+namespace mbed {
+/** \addtogroup drivers */
+
+/** CPU UID reader class
+ *
+ * @note Synchronization level: Interrupt safe
```

Comment:
Is this class really interrupt safe? For instance methods like `operator std::string()` ?

Reply:
I'm not sure ... did not think much about it - just copy&pasted the doxygen blocks from another driver ;)

Reply:
why is STL not interrupt safe?

Reply:
Because of memory allocation. Calling malloc in an IRQ can result in a deadlock.

Reply:
and what if memory allocation is secured by a CriticalSectionLock?
I am asking because I have written another class for our project, which does allocate buffers within an interrupt callback function ... I secured it with a critical section lock, but now I am not completely sure if this is enough ;)

Reply:
If a memory allocation is secured with a mutex, and an allocation is in progress when an IRQ fires, and that IRQ also allocates memory, then the IRQ will wait forever because it is not possible to switch into the in-progress thread in order to free up the mutex.

If memory allocation is secured with a critical section, deadlocks are eliminated, but IRQ latency and processing time become unpredictable.

IRQs are best used for signaling events to a main loop or thread.

Reply:
Ok that means if I only use a critical section to be interrupt-safe, I won't be thread safe?

In particular, that other class of mine makes use of blocking and non-blocking UART communication.
I use the critical section to append a log-entry to a queue which gets processed asynchronously ... which is used for logging purposes within irq handler functions.
The blocking part is used for non-irq functions.

So if we should decide to make use of threading in our project, I should also use mutexes for thread safety?
Or is the Serial class already thread safe?

Reply:
Critical sections are thread safe. The problem is that entering and exiting a critical section is expensive on some platforms. It also means that you disable context switching, not just interrupts. Furthermore, because a critical section prevents interrupts, it may break assumptions about interrupt handling latency.

In mbed, with gcc, we use newlib. newlib's approach to handling re-entrancy in malloc is to call a function called `__malloc_lock()` to acquire an exclusive lock on `malloc`. It also calls `__malloc_unlock()` to release `malloc`. In [mbed-retarget.cpp](https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_retarget.cpp#L960) this is defined as a call-through to `__rtos_malloc_lock()`/`__rtos_malloc_unlock()`. These, in turn, are defined in [mbed_boot.c](https://github.com/ARMmbed/mbed-os/blob/master/rtos/TARGET_CORTEX/mbed_boot.c#L578). `__rtos_malloc_lock()` simply calles `osMutexAcquire`.

`malloc()` is called in many places, including `printf()`, `new` and other deeply nested APIs. It is not practical to wrap all of them with `CriticalSectionLock()`. If any of those calls to `malloc` are in progress when an IRQ fires, and calls `malloc()`, it doesn't matter whether or not there is a critical section around the call to `malloc()`: the call that is already in progress still holds the mutex and it is not possible to context switch out of an interrupt. This will create a deadlock.

Therefore, it is important to never call `malloc()` or `new` or `free` in an IRQ. These must be deferred to thread context using a semaphore, flag variable, or other synchronization mechanism.